File: frontend/src/services/postService.js
import api from './api';

class PostService {
  async getApprovedPosts(page = 0, size = 10) {
    const response = await api.get('/posts', { params: { page, size } });
    return response.data;
  }

  async getApprovedPostsByCategory(categoryId, page = 0, size = 10) {
    const response = await api.get('/posts', { params: { categoryId, page, size } });
    return response.data;
  }

  async getPostById(id) {
    const response = await api.get(`/posts/${id}`);
    return response.data;
  }

  async createPost(data) {
    const response = await api.post('/author/posts', data);
    return response.data;
  }

  async updatePost(id, data) {
    const response = await api.put(`/author/posts/${id}`, data);
    return response.data;
  }

  async deletePost(id) {
    const response = await api.delete(`/author/posts/${id}`);
    return response.data;
  }

  async submitForReview(id) {
    const response = await api.post(`/author/posts/${id}/submit`);
    return response.data;
  }

  async getMyPosts(page = 0, size = 10) {
    const response = await api.get('/author/posts/my', { params: { page, size } });
    return response.data;
  }

  async getMyPostsByStatus(status) {
    const response = await api.get(`/author/posts/my/status/${status}`);
    return response.data;
  }

  async toggleLike(postId) {
    const response = await api.post(`/posts/${postId}/like`);
    return response.data;
  }

  async getComments(postId) {
    const response = await api.get(`/posts/${postId}/comments`);
    return response.data;
  }

  async addComment(postId, content, parentCommentId = null) {
    const response = await api.post(`/posts/${postId}/comments`, { 
      content,
      parentCommentId 
    });
    return response.data;
  }

  // Report post
  async reportPost(postId, reason, description) {
    const response = await api.post(`/posts/${postId}/report`, { reason, description });
    return response.data;
  }

  // Admin endpoints
  async getPendingPosts(page = 0, size = 10) {
    const response = await api.get('/admin/posts/pending', { params: { page, size } });
    return response.data;
  }

  async getReviewPosts(page = 0, size = 10) {
    const response = await api.get('/admin/posts/review', { params: { page, size } });
    return response.data;
  }

  async approvePost(id, note) {
    const response = await api.post(`/admin/posts/${id}/approve`, { note });
    return response.data;
  }

  async rejectPost(id, note) {
    const response = await api.post(`/admin/posts/${id}/reject`, { note });
    return response.data;
  }

  // Admin Report Management
  async getAllReports(page = 0, size = 10) {
    const response = await api.get('/admin/reports', { params: { page, size } });
    return response.data;
  }

  async getPendingReports(page = 0, size = 10) {
    const response = await api.get('/admin/reports/pending', { params: { page, size } });
    return response.data;
  }

  async reviewReport(id, status, note) {
    const response = await api.post(`/admin/reports/${id}/review`, { note }, { params: { status } });
    return response.data;
  }
}

export default new PostService();

File: backend/src/main/java/com/medee/service/PostService.java
package com.medee.service;

import com.medee.dto.PostDto;
import com.medee.dto.PostRequest;
import com.medee.model.Category;
import com.medee.model.Post;
import com.medee.model.PostStatus;
import com.medee.model.User;
import com.medee.repository.CategoryRepository;
import com.medee.repository.PostRepository;
import com.medee.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class PostService {

    private final PostRepository postRepository;
    private final UserRepository userRepository;
    private final CategoryRepository categoryRepository;

    public PostDto createPost(String authorId, PostRequest request) {
        Post post = Post.builder()
                .title(request.getTitle())
                .content(request.getContent())
                .authorId(authorId)
                .categoryId(request.getCategoryId())
                .status(PostStatus.DRAFT)
                .likes(0)
                .views(0)
                .build();

        post = postRepository.save(post);
        return mapToDto(post);
    }

    public PostDto updatePost(String postId, String authorId, PostRequest request) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (!post.getAuthorId().equals(authorId)) {
            throw new RuntimeException("You are not authorized to update this post");
        }

        post.setTitle(request.getTitle());
        post.setContent(request.getContent());
        if (request.getCategoryId() != null) {
            post.setCategoryId(request.getCategoryId());
        }
        post = postRepository.save(post);

        return mapToDto(post);
    }

    public void deletePost(String postId, String authorId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (!post.getAuthorId().equals(authorId)) {
            throw new RuntimeException("You are not authorized to delete this post");
        }

        postRepository.delete(post);
    }

    public PostDto submitForReview(String postId, String authorId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (!post.getAuthorId().equals(authorId)) {
            throw new RuntimeException("You are not authorized to submit this post");
        }

        if (post.getStatus() != PostStatus.DRAFT && post.getStatus() != PostStatus.REJECTED) {
            throw new RuntimeException("Only draft or rejected posts can be submitted for review");
        }

        post.setStatus(PostStatus.PENDING_REVIEW);
        post = postRepository.save(post);

        return mapToDto(post);
    }

    public Page<PostDto> getMyPosts(String authorId, Pageable pageable) {
        Page<Post> posts = postRepository.findByAuthorId(authorId, pageable);
        return posts.map(this::mapToDto);
    }

    public List<PostDto> getMyPostsByStatus(String authorId, PostStatus status) {
        List<Post> posts = postRepository.findByAuthorIdAndStatus(authorId, status);
        return posts.stream().map(this::mapToDto).collect(Collectors.toList());
    }

    public Page<PostDto> getApprovedPosts(Pageable pageable) {
        Page<Post> posts = postRepository.findByStatus(PostStatus.APPROVED, pageable);
        return posts.map(this::mapToDto);
    }

    public Page<PostDto> getApprovedPostsByCategory(String categoryId, Pageable pageable) {
        Page<Post> posts = postRepository.findByCategoryIdAndStatus(categoryId, PostStatus.APPROVED, pageable);
        return posts.map(this::mapToDto);
    }

    public PostDto getPostById(String postId, boolean incrementView) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (incrementView) {
            post.setViews(post.getViews() + 1);
            postRepository.save(post);
        }

        return mapToDto(post);
    }

    public PostDto approvePost(String postId, String adminId, String note) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (post.getStatus() != PostStatus.PENDING_REVIEW) {
            throw new RuntimeException("Only pending posts can be approved");
        }

        post.setStatus(PostStatus.APPROVED);
        post.setReviewedBy(adminId);
        post.setReviewNote(note);
        post.setPublishedAt(LocalDateTime.now());
        post = postRepository.save(post);

        return mapToDto(post);
    }

    public PostDto rejectPost(String postId, String adminId, String note) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (post.getStatus() != PostStatus.PENDING_REVIEW) {
            throw new RuntimeException("Only pending posts can be rejected");
        }

        post.setStatus(PostStatus.REJECTED);
        post.setReviewedBy(adminId);
        post.setReviewNote(note);
        post = postRepository.save(post);

        return mapToDto(post);
    }

    public Page<PostDto> getPendingPosts(Pageable pageable) {
        Page<Post> posts = postRepository.findByStatus(PostStatus.PENDING_REVIEW, pageable);
        return posts.map(this::mapToDto);
    }

    public Page<PostDto> getReviewPosts(String adminId, Pageable pageable) {
        Page<Post> posts = postRepository.findByAuthorIdNotAndStatusNot(
                adminId,
                PostStatus.DRAFT,
                pageable
        );
        return posts.map(this::mapToDto);
    }

    private PostDto mapToDto(Post post) {
        User author = userRepository.findById(post.getAuthorId()).orElse(null);
        String authorName = author != null && author.getProfile() != null 
                ? author.getProfile().getDisplayName() 
                : (author != null ? author.getUsername() : "Unknown");

        String categoryName = null;
        if (post.getCategoryId() != null) {
            Category category = categoryRepository.findById(post.getCategoryId()).orElse(null);
            if (category != null) {
                categoryName = category.getName();
            }
        }

        return PostDto.builder()
                .id(post.getId())
                .title(post.getTitle())
                .content(post.getContent())
                .authorId(post.getAuthorId())
                .authorName(authorName)
                .categoryId(post.getCategoryId())
                .categoryName(categoryName)
                .status(post.getStatus())
                .likes(post.getLikes())
                .views(post.getViews())
                .createdAt(post.getCreatedAt())
                .updatedAt(post.getUpdatedAt())
                .publishedAt(post.getPublishedAt())
                .reviewNote(post.getReviewNote())
                .build();
    }
}

